plugins/outputs/whatap/whatap.go

```diff
@@ -0,0 +1,382 @@
+	Oname   string
+	Oid     int32
+	Session TcpSession
+}
+
+type TcpSession struct {
+	Client net.Conn
+	Dest   int
+}
+
+var sampleConfig = `
+  ## You can create a project on the WhaTap site(https://www.whatap.io) 
+  ## to get license, project code and server IP information.
+
+  ## WhaTap license. Required
+  #license = "xxxx-xxxx-xxxx"
+
+  ## WhaTap project code. Required
+  #pcode = 1111
+
+  ## WhaTap server IP. Required
+  ## Put multiple IPs. ["tcp://1.1.1.1:6600","tcp://2.2.2.2:6600"]
+  #servers = ["tcp://1.1.1.1:6600","tcp://2.2.2.2:6600"]
+
+  ## Connection timeout.
+  # timeout = "60s"
+`
+
+func (w *Whatap) Connect() error {
+	//log.Println("[outputs.whatap] Connect")
+	w.Session.Dest += 1
+	if w.Session.Dest >= len(w.Servers) {
+		w.Session.Dest = 0
+	}
+
+	addr := strings.SplitN(w.Servers[w.Session.Dest], "://", 2)
+	if len(addr) != 2 {
+		return fmt.Errorf("invalid address: %s", w.Servers[w.Session.Dest])
+	}
+	if addr[0] != "tcp" {
+		return fmt.Errorf("only tcp is supported: %s", w.Servers[w.Session.Dest])
+	}
+
+	t := w.Timeout * time.Millisecond
+	if client, err := net.DialTimeout(addr[0], addr[1], t); err != nil {
+		return err
+	} else {
+		w.Session.Client = client.(*net.TCPConn)
+		log.Println("I! [outputs.whatap] Connected tcp to ", addr)
+	}
+
+	return nil
+
+}
+
+func (w *Whatap) Close() error {
+	if w.Session.Client == nil {
+		return nil
+	}
+	err := w.Session.Client.Close()
+	w.Session.Client = nil
+	return err
+}
+
+func (w *Whatap) Description() string {
+	return "Configuration for WhaTap"
+}
+
+func (w *Whatap) SampleConfig() string {
+	return sampleConfig
+}
+
+func (w *Whatap) Write(metrics []telegraf.Metric) error {
+	if len(metrics) == 0 {
+		return nil
+	}
+	if w.Session.Client == nil {
+		if err := w.Connect(); err != nil {
+			w.Close()
+			return err
+		}
+	}
+	for _, m := range metrics {
+		p := NewTagCountPack()
+		p.Pcode = w.Pcode
+		p.Oid = w.Oid
+		p.Category = fmt.Sprintf("%s%s", "telegraf_", m.Name())
+		for k, v := range m.Fields() {
+			p.Put(k, v)
+		}
+		for k, v := range m.Tags() {
+			p.PutTag(k, v)
+		}
+		// Add Oname
+		p.PutTag("oname", w.Oname)
+		p.Time = m.Time().UnixNano() / int64(time.Millisecond)
+
+		dout := NewDataOutputX()
+		// pack type
+		dout.WriteShort(p.GetPackType())
+		p.Write(dout)
+
+		if err := w.send(0, dout.ToByteArray()); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+func (w *Whatap) send(code byte, b []byte) (err error) {
+	out := NewDataOutputX()
+	out.WriteByte1(NETSRC_AGENT_ONEWAY)
+	// ver
+	out.WriteByte1(code)
+	out.WriteLong(w.Pcode)
+	out.WriteLong(Hash64Str(w.License))
+	out.WriteIntBytes(b)
+	// pack data
+	sendbuf := out.ToByteArray()
+
+	nbyteleft := len(sendbuf)
+	pos := 0
+	for 0 < nbyteleft {
+		nbytethistime := 0
+		// Set Deadline
+		err = w.Session.Client.SetWriteDeadline(time.Now().Add(
+			w.Timeout * time.Millisecond))
+		if err != nil {
+			log.Println("W! [outputs.whatap] cannot set tcp write deadline:",
+				err)
+		}
```

Comment:
I think this counts for a non-recoverable error
```suggestion
		deadline := time.Now().Add(w.Timeout)
		if err := w.Session.Client.SetWriteDeadline(deadline);  err != nil {
			return fmt.Errorf("cannot set write deadline: %v", err)
		}
```

Reply:
Edited and applied. We ask for confirmation.